<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Data visualization • TRONCO</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Data visualization">
<meta property="og:description" content="TRONCO">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TRONCO</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="Released version">2.24.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/1_introduction.html">Introduction</a>
    </li>
    <li>
      <a href="../articles/2_loading_data.html">Loading Data</a>
    </li>
    <li>
      <a href="../articles/3_data_visualization.html">Data visualization</a>
    </li>
    <li>
      <a href="../articles/4_data_manipulation.html">Data manipulation</a>
    </li>
    <li>
      <a href="../articles/5_model_inference.html">Model inference</a>
    </li>
    <li>
      <a href="../articles/6_post_reconstruction.html">Post-reconstruction</a>
    </li>
    <li>
      <a href="../articles/7_import_export.html">Import/export from other tools</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BIMIB-DISCo/TRONCO/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><script src="3_data_visualization_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Data visualization</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/BIMIB-DISCo/TRONCO/blob/master/vignettes/3_data_visualization.Rmd" class="external-link"><code>vignettes/3_data_visualization.Rmd</code></a></small>
      <div class="hidden name"><code>3_data_visualization.Rmd</code></div>

    </div>

    
    
<p>All examples in this section will be done with the the aCML dataset as reference.</p>
<div id="sec:view" class="section level2">
<h2 class="hasAnchor">
<a href="#sec:view" class="anchor"></a>Summary report for a dataset and boolean queries</h2>
<p>We use the function <code>view</code> to get a short summary of a dataset that we loaded in TRONCO; this function reports on the number of samples and events, plus some meta information that could be displayed graphically.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/view.html">view</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span></code></pre></div>
<pre><code>## Description: CAPRI - Bionformatics aCML data.
## -- TRONCO Dataset: n=64, m=31, |G|=23, patterns=0.
## Events (types): Ins/Del, Missense point, Nonsense Ins/Del, Nonsense point.
## Colors (plot): #7FC97F, #4483B0, #FDC086, #fab3d8.
## Events (5 shown):
##   gene 4 : Ins/Del TET2
##   gene 5 : Ins/Del EZH2
##   gene 6 : Ins/Del CBL
##   gene 7 : Ins/Del ASXL1
##   gene 29 : Missense point SETBP1
## Genotypes (5 shown):</code></pre>
</div>
<div id="creating-views-with-the-as-functions" class="section level2">
<h2 class="hasAnchor">
<a href="#creating-views-with-the-as-functions" class="anchor"></a>Creating views with the <code>as</code> functions</h2>
<p>Several functions are available to create views over a dataset, with a set of parameter which can constraint the view – as in the SELECT/JOIN approaches in databases. In the following examples we show their execution with the default parameters, but shorten their output to make this document readable.</p>
<p>The main <code>as</code> functions are here documented. <code>as.genotypes</code>, that we can use to get the matrix of <em>genotypes</em> that we imported.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/as.genotypes.html">as.genotypes</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span>,<span class="fl">5</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></code></pre></div>
<pre><code>##            gene 29 gene 30 gene 31 gene 32 gene 33 gene 34
## patient 1        1       0       0       0       0       0
## patient 2        1       0       0       0       0       1
## patient 3        1       1       0       0       0       0
## patient 4        1       0       0       0       0       1
## patient 5        1       0       0       0       0       0
## patient 6        1       0       0       0       0       0
## patient 7        1       0       0       0       0       0
## patient 8        1       0       0       0       0       0
## patient 9        1       0       0       0       0       0
## patient 10       1       0       0       0       0       0</code></pre>
<p>Differently, <code>as.events</code> and <code>as.events.in.samples</code>, that show tables with the events that we are processing in all dataset or in a specific sample that we want to examine.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/as.events.html">as.events</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">5</span>, <span class="op">]</span></code></pre></div>
<pre><code>##         type             event   
## gene 4  "Ins/Del"        "TET2"  
## gene 5  "Ins/Del"        "EZH2"  
## gene 6  "Ins/Del"        "CBL"   
## gene 7  "Ins/Del"        "ASXL1" 
## gene 29 "Missense point" "SETBP1"</code></pre>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/as.events.in.sample.html">as.events.in.sample</a></span><span class="op">(</span><span class="va">aCML</span>, sample <span class="op">=</span> <span class="st">'patient 2'</span><span class="op">)</span></code></pre></div>
<pre><code>##         type             event   
## gene 29 "Missense point" "SETBP1"
## gene 34 "Missense point" "CBL"   
## gene 91 "Nonsense point" "ASXL1"</code></pre>
<p>Concerning genes, <code>as.genes</code> shows the mnemonic names of the genes (or chromosomes, cytobands, etc.) that we included in our dataset.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/as.genes.html">as.genes</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">8</span><span class="op">]</span></code></pre></div>
<pre><code>## [1] "TET2"   "EZH2"   "CBL"    "ASXL1"  "SETBP1" "NRAS"   "KRAS"   "IDH2"</code></pre>
<p>And <code>as.types</code> shows the types of alterations (e.g., mutations, amplifications, etc.) that we have find in our dataset, and function <code>as.colors</code> shows the list of the colors which are associated to each type.</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/as.types.html">as.types</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "Ins/Del"          "Missense point"   "Nonsense Ins/Del" "Nonsense point"</code></pre>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/as.colors.html">as.colors</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span></code></pre></div>
<pre><code>##          Ins/Del   Missense point Nonsense Ins/Del   Nonsense point 
##        "#7FC97F"        "#4483B0"        "#FDC086"        "#fab3d8"</code></pre>
<p>A function <code>as.gene</code> can be used to display the alterations of a specific gene across the samples</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/as.gene.html">as.gene</a></span><span class="op">(</span><span class="va">aCML</span>, genes<span class="op">=</span><span class="st">'SETBP1'</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##           Missense point SETBP1
## patient 1                     1
## patient 2                     1
## patient 3                     1
## patient 4                     1
## patient 5                     1
## patient 6                     1</code></pre>
<p>Views over samples can be created as well. <code>as.samples</code> and <code>which.samples</code> list all the samples in the data, or return a list of samples that harbour a certain alteration. The former is</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/as.samples.html">as.samples</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span></code></pre></div>
<pre><code>##  [1] "patient 1"  "patient 2"  "patient 3"  "patient 4"  "patient 5" 
##  [6] "patient 6"  "patient 7"  "patient 8"  "patient 9"  "patient 10"</code></pre>
<p>and the latter is</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/which.samples.html">which.samples</a></span><span class="op">(</span><span class="va">aCML</span>, gene<span class="op">=</span><span class="st">'TET2'</span>, type<span class="op">=</span><span class="st">'Nonsense point'</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] "patient 12" "patient 13" "patient 26" "patient 29" "patient 40"
## [6] "patient 57" "patient 62"</code></pre>
<p>A slightly different function, which manipulates the data, is <code>as.alterations</code>, which transforms a dataset with events of different type to events of a unique type, labeled <em>Alteration</em>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">dataset</span> <span class="op">=</span> <span class="fu"><a href="../reference/as.alterations.html">as.alterations</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span></code></pre></div>
<pre><code>## *** Aggregating events of type(s) { Ins/Del, Missense point, Nonsense Ins/Del, Nonsense point }
## in a unique event with label " Alteration ".
## Dropping event types Ins/Del, Missense point, Nonsense Ins/Del, Nonsense point for 23 genes.
## .......................
## *** Binding events for 2 datasets.</code></pre>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/view.html">view</a></span><span class="op">(</span><span class="va">dataset</span><span class="op">)</span></code></pre></div>
<pre><code>## -- TRONCO Dataset: n=64, m=23, |G|=23, patterns=0.
## Events (types): Alteration.
## Colors (plot): khaki.
## Events (5 shown):
##   G1 : Alteration TET2
##   G2 : Alteration EZH2
##   G3 : Alteration CBL
##   G4 : Alteration ASXL1
##   G5 : Alteration SETBP1
## Genotypes (5 shown):</code></pre>
<p>When samples are enriched with stage information function <code>as.stages</code> can be used to create a view over such table. Views over patterns can be created as well – see Model Inference with CAPRI.</p>
</div>
<div id="dataset-size" class="section level2">
<h2 class="hasAnchor">
<a href="#dataset-size" class="anchor"></a>Dataset size</h2>
<p>A set of functions allow to get the number of genes, events, samples, types and patterns in a dataset.</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/ngenes.html">ngenes</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 23</code></pre>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/nevents.html">nevents</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 31</code></pre>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/nsamples.html">nsamples</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 64</code></pre>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/ntypes.html">ntypes</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 4</code></pre>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/npatterns.html">npatterns</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] 0</code></pre>
</div>
<div id="oncoprints" class="section level2">
<h2 class="hasAnchor">
<a href="#oncoprints" class="anchor"></a>Oncoprints</h2>
<p>Oncoprints are the most effective data-visualization functions in TRONCO. These are heatmaps where rows represent variants, and columns samples (<em>the reverse</em> of the input format required by TRONCO), and are annotated and displayed/sorted to enhance which samples have which mutations etc.</p>
<p>By default <code>oncoprint</code> will try to sort samples and events to enhance exclusivity patterns among the events.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/oncoprint.html">oncoprint</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span></code></pre></div>
<pre><code>## *** Oncoprint for "CAPRI - Bionformatics aCML data"
## with attributes: stage = FALSE, hits = TRUE
## Sorting samples ordering to enhance exclusivity patterns.
## Setting automatic row font (exponential scaling): 8.1</code></pre>
<div class="figure">
<img src="3_data_visualization_files/figure-html/unnamed-chunk-13-1.png" alt="This plot gives a graphical visualization of the events that are in the dataset -- with a color per event type. It sorts samples to enhance exclusivity patterns among the events" width="576"><p class="caption">
This plot gives a graphical visualization of the events that are in the dataset – with a color per event type. It sorts samples to enhance exclusivity patterns among the events
</p>
</div>
<p>But the sorting mechanism is bypassed if one wants to cluster samples or events, or if one wants to split samples by cluster (not shown). In the clustering case, the ordering is given by the dendrograms. In this case we also show the annotation of some groups of events via parameter <code>gene.annot</code>.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/oncoprint.html">oncoprint</a></span><span class="op">(</span><span class="va">aCML</span>, 
    legend <span class="op">=</span> <span class="cn">FALSE</span>, 
    samples.cluster <span class="op">=</span> <span class="cn">TRUE</span>, 
    gene.annot <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>one <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'NRAS'</span>, <span class="st">'SETBP1'</span><span class="op">)</span>, two <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="st">'EZH2'</span>, <span class="st">'TET2'</span><span class="op">)</span><span class="op">)</span>,
    gene.annot.color <span class="op">=</span> <span class="st">'Set2'</span>,
    genes.cluster <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></code></pre></div>
<pre><code>## *** Oncoprint for "CAPRI - Bionformatics aCML data"
## with attributes: stage = FALSE, hits = TRUE
## Sorting samples ordering to enhance exclusivity patterns.
## Annotating genes with RColorBrewer color palette Set2 .
## Setting automatic row font (exponential scaling): 8.1
## Clustering samples and showing dendogram.
## Clustering alterations and showing dendogram.</code></pre>
<div class="figure">
<img src="3_data_visualization_files/figure-html/unnamed-chunk-14-1.png" alt="This plot gives a graphical visualization of the events that are in the dataset -- with a color per event type. It it clusters samples/events" width="480"><p class="caption">
This plot gives a graphical visualization of the events that are in the dataset – with a color per event type. It it clusters samples/events
</p>
</div>
<p>Oncoprints can be annotated; a special type of annotation is given by stage data. As this is not available for the aCML dataset, we create it randomly, just for the sake of showing how the oncoprint is enriched with this information. This is the random stage map that we create – if some samples had no stage a NA would be added automatically.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">stages</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'stage 1'</span>, <span class="fl">32</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="st">'stage 2'</span>, <span class="fl">32</span><span class="op">)</span><span class="op">)</span>
<span class="va">stages</span> <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/matrix.html" class="external-link">as.matrix</a></span><span class="op">(</span><span class="va">stages</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">stages</span><span class="op">)</span> <span class="op">=</span> <span class="fu"><a href="../reference/as.samples.html">as.samples</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span>
<span class="va">dataset</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate.stages.html">annotate.stages</a></span><span class="op">(</span><span class="va">aCML</span>, stages <span class="op">=</span> <span class="va">stages</span><span class="op">)</span>
<span class="fu"><a href="../reference/has.stages.html">has.stages</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span></code></pre></div>
<pre><code>## [1] FALSE</code></pre>
<div class="sourceCode" id="cb41"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/as.stages.html">as.stages</a></span><span class="op">(</span><span class="va">dataset</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##             stage
## patient 1 stage 1
## patient 2 stage 1
## patient 3 stage 1
## patient 4 stage 1
## patient 5 stage 1
## patient 6 stage 1</code></pre>
<p>The <code>as.stages</code> function can now be used to create a view over stages.</p>
<div class="sourceCode" id="cb43"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="fu"><a href="../reference/as.stages.html">as.stages</a></span><span class="op">(</span><span class="va">dataset</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>##             stage
## patient 1 stage 1
## patient 2 stage 1
## patient 3 stage 1
## patient 4 stage 1
## patient 5 stage 1
## patient 6 stage 1</code></pre>
<p>After that the data is annotated via <code>annotate.stages</code> function, we can again plot an oncoprint – which this time will detect that the dataset has also stages associated, and will diplay those</p>
<div class="sourceCode" id="cb45"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/oncoprint.html">oncoprint</a></span><span class="op">(</span><span class="va">dataset</span>, legend <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code>## *** Oncoprint for "CAPRI - Bionformatics aCML data"
## with attributes: stage = TRUE, hits = TRUE
## Sorting samples ordering to enhance exclusivity patterns.
## Annotating stages with RColorBrewer color palette YlOrRd 
## Setting automatic row font (exponential scaling): 8.1</code></pre>
<p><img src="3_data_visualization_files/figure-html/unnamed-chunk-17-1.png" width="576"></p>
<p>If one is willing to display samples grouped according to some variable, for instance after a sample clustering task, he can use <code>group.samples</code> parameter of <code>oncoprint</code> and that will override the mutual exclusivity ordering. Here, we make the trick of using the stages as if they were such clustering result.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/oncoprint.html">oncoprint</a></span><span class="op">(</span><span class="va">dataset</span>, group.samples <span class="op">=</span> <span class="fu"><a href="../reference/as.stages.html">as.stages</a></span><span class="op">(</span><span class="va">dataset</span><span class="op">)</span><span class="op">)</span></code></pre></div>
<pre><code>## *** Oncoprint for "CAPRI - Bionformatics aCML data"
## with attributes: stage = TRUE, hits = TRUE
## Sorting samples ordering to enhance exclusivity patterns.
## Grouping samples according to input groups (group.samples).
## Annotating stages with RColorBrewer color palette YlOrRd 
## Grouping labels: stage 1, stage 2 
## Setting automatic row font (exponential scaling): 8.1</code></pre>
<div class="figure">
<img src="3_data_visualization_files/figure-html/unnamed-chunk-18-1.png" alt="Example    exttt{oncoprint} output for aCML data with randomly annotated stages, in left, and samples clustered by group assignment in right -- for simplicity the group variable is again the stage annotation." width="576"><p class="caption">
Example exttt{oncoprint} output for aCML data with randomly annotated stages, in left, and samples clustered by group assignment in right – for simplicity the group variable is again the stage annotation.
</p>
</div>
</div>
<div id="groups-visualization-e-g--pathways" class="section level2">
<h2 class="hasAnchor">
<a href="#groups-visualization-e-g--pathways" class="anchor"></a>Groups visualization (e.g., pathways)</h2>
<p>TRONCO provides functions to visualize groups of events, which in this case are called pathways – though this could be any group that one would like to define. Aggregation happens with the same rational as the <code>as.alterations</code> function, namely by merging the events in the group.</p>
<p>We make an example of a pathway called <em>MyPATHWAY</em> involving genes SETBP1, EZH2 and WT1; we want it to be colored in red, and we want to have the genotype of each event to be maintened in the dataset. We proceed as follows (R’s output is omitted).</p>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="va">pathway</span> <span class="op">=</span> <span class="fu"><a href="../reference/as.pathway.html">as.pathway</a></span><span class="op">(</span><span class="va">aCML</span>,
    pathway.genes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SETBP1'</span>, <span class="st">'EZH2'</span>, <span class="st">'WT1'</span><span class="op">)</span>,
    pathway.name <span class="op">=</span> <span class="st">'MyPATHWAY'</span>,
    pathway.color <span class="op">=</span> <span class="st">'red'</span>,
    aggregate.pathway <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></code></pre></div>
<pre><code>## *** Extracting events for pathway:  MyPATHWAY .
## *** Events selection: #events =  31 , #types =  4 Filters freq|in|out = { FALSE ,  TRUE ,  FALSE }
## [filter.in] Genes hold:  SETBP1, EZH2, WT1  ...  [ 3 / 3  found].
## Selected  5  events, returning.
## Pathway extracted succesfully.
## *** Binding events for 2 datasets.</code></pre>
<p>Which we then visualize with an <code>oncoprint</code></p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/oncoprint.html">oncoprint</a></span><span class="op">(</span><span class="va">pathway</span>, title <span class="op">=</span> <span class="st">'Custom pathway'</span>,  font.row <span class="op">=</span> <span class="fl">8</span>, cellheight <span class="op">=</span> <span class="fl">15</span>, cellwidth <span class="op">=</span> <span class="fl">4</span><span class="op">)</span></code></pre></div>
<pre><code>## *** Oncoprint for "Custom pathway"
## with attributes: stage = FALSE, hits = TRUE
## Sorting samples ordering to enhance exclusivity patterns.</code></pre>
<div class="figure">
<img src="3_data_visualization_files/figure-html/onco-pathway-1.png" alt="Oncoprint output of a custom pathway called MyPATHWAY involving genes SETBP1, EZH2 and WT1; the genotype of each event is shown." width="624"><p class="caption">
Oncoprint output of a custom pathway called MyPATHWAY involving genes SETBP1, EZH2 and WT1; the genotype of each event is shown.
</p>
</div>
<p>In TRONCO there is also a function which creates the pathway view and the corresponding oncoprint to multiple pathways, when these are given as a list. We make here a simple example of two custom pathways.</p>
<div class="sourceCode" id="cb53"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/pathway.visualization.html">pathway.visualization</a></span><span class="op">(</span><span class="va">aCML</span>, 
    pathways<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>P1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'TET2'</span>, <span class="st">'IRAK4'</span><span class="op">)</span>,  P2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SETBP1'</span>, <span class="st">'KIT'</span><span class="op">)</span><span class="op">)</span>,        
    aggregate.pathways<span class="op">=</span><span class="cn">FALSE</span>,
    font.row <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></code></pre></div>
<pre><code>## Annotating pathways with RColorBrewer color palette Set2 .
## *** Processing pathways: P1, P2
## 
## [PATHWAY "P1"] TET2, IRAK4
## *** Extracting events for pathway:  P1 .
## *** Events selection: #events =  31 , #types =  4 Filters freq|in|out = { FALSE ,  TRUE ,  FALSE }
## [filter.in] Genes hold:  TET2, IRAK4  ...  [ 2 / 2  found].
## Selected  4  events, returning.
## Pathway extracted succesfully.
## *** Binding events for 2 datasets.
## 
## 
## [PATHWAY "P2"] SETBP1, KIT
## *** Extracting events for pathway:  P2 .
## *** Events selection: #events =  31 , #types =  4 Filters freq|in|out = { FALSE ,  TRUE ,  FALSE }
## [filter.in] Genes hold:  SETBP1, KIT  ...  [ 2 / 2  found].
## Selected  2  events, returning.
## Pathway extracted succesfully.
## *** Binding events for 2 datasets.
## *** Binding events for 2 datasets.
## *** Oncoprint for "Pathways: P1, P2"
## with attributes: stage = FALSE, hits = TRUE
## Sorting samples ordering to enhance exclusivity patterns.</code></pre>
<div class="figure">
<img src="3_data_visualization_files/figure-html/unnamed-chunk-20-1.png" alt="Oncoprint output of a custom pair of pathways, with events shown" width="624"><p class="caption">
Oncoprint output of a custom pair of pathways, with events shown
</p>
</div>
<pre><code>## NULL</code></pre>
<p>If we had to visualize just the signature of the pathway, we could set <code>aggregate.pathways=T</code>.</p>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="../reference/pathway.visualization.html">pathway.visualization</a></span><span class="op">(</span><span class="va">aCML</span>, 
    pathways<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>P1 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'TET2'</span>, <span class="st">'IRAK4'</span><span class="op">)</span>,  P2<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'SETBP1'</span>, <span class="st">'KIT'</span><span class="op">)</span><span class="op">)</span>,
    aggregate.pathways <span class="op">=</span> <span class="cn">TRUE</span>,
    font.row <span class="op">=</span> <span class="fl">8</span><span class="op">)</span></code></pre></div>
<pre><code>## Annotating pathways with RColorBrewer color palette Set2 .
## *** Processing pathways: P1, P2
## 
## [PATHWAY "P1"] TET2, IRAK4
## *** Extracting events for pathway:  P1 .
## *** Events selection: #events =  31 , #types =  4 Filters freq|in|out = { FALSE ,  TRUE ,  FALSE }
## [filter.in] Genes hold:  TET2, IRAK4  ...  [ 2 / 2  found].
## Selected  4  events, returning.
## Pathway extracted succesfully.
## 
## 
## [PATHWAY "P2"] SETBP1, KIT
## *** Extracting events for pathway:  P2 .
## *** Events selection: #events =  31 , #types =  4 Filters freq|in|out = { FALSE ,  TRUE ,  FALSE }
## [filter.in] Genes hold:  SETBP1, KIT  ...  [ 2 / 2  found].
## Selected  2  events, returning.
## Pathway extracted succesfully.
## *** Binding events for 2 datasets.
## *** Oncoprint for "Pathways: P1, P2"
## with attributes: stage = FALSE, hits = TRUE
## Sorting samples ordering to enhance exclusivity patterns.</code></pre>
<div class="figure">
<img src="3_data_visualization_files/figure-html/unnamed-chunk-21-1.png" alt="Oncoprint output of a custom pair of pathways, with events hidden" width="624"><p class="caption">
Oncoprint output of a custom pair of pathways, with events hidden
</p>
</div>
<pre><code>## NULL</code></pre>
<p>The same operation could have been done using <a href="https://wikipathways.org" class="external-link">WikiPathways</a>. We can query WikiPathways and collect HGNC gene symbols and titles for pathways of interest as follows. (R’s output is omitted).</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/wikipathways/rwikipathways" class="external-link">rWikiPathways</a></span><span class="op">)</span>
<span class="co"># quotes inside query to require both terms</span>
<span class="va">my.pathways</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rWikiPathways/man/findPathwaysByText.html" class="external-link">findPathwaysByText</a></span><span class="op">(</span><span class="st">'SETBP1 EZH2 TET2 IRAK4 SETBP1 KIT'</span><span class="op">)</span>
<span class="va">human.filter</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">lapply</a></span><span class="op">(</span><span class="va">my.pathways</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">species</span> <span class="op">==</span> <span class="st">"Homo sapiens"</span><span class="op">)</span>
<span class="va">my.hs.pathways</span> <span class="op">&lt;-</span> <span class="va">my.pathways</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/unlist.html" class="external-link">unlist</a></span><span class="op">(</span><span class="va">human.filter</span><span class="op">)</span><span class="op">]</span> 
<span class="co"># collect pathways idenifiers</span>
<span class="va">my.wpids</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/lapply.html" class="external-link">sapply</a></span><span class="op">(</span><span class="va">my.hs.pathways</span>, <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span> <span class="va">x</span><span class="op">$</span><span class="va">id</span><span class="op">)</span>

<span class="va">pw.title</span><span class="op">&lt;-</span><span class="va">my.hs.pathways</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span><span class="op">$</span><span class="va">name</span>
<span class="va">pw.genes</span><span class="op">&lt;-</span><span class="fu"><a href="https://rdrr.io/pkg/rWikiPathways/man/getXrefList.html" class="external-link">getXrefList</a></span><span class="op">(</span><span class="va">my.wpids</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span>,<span class="st">"H"</span><span class="op">)</span> </code></pre></div>
<p>Now <code>pw.genes</code> and <code>pw.title</code> can be used as input for the function <code>as.pathway</code>. It is also possible to view and edit these pathways at WikiPathways using the following commands to open tabs in your default browser.</p>
<div class="sourceCode" id="cb60"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span class="fu"><a href="https://rdrr.io/r/utils/browseURL.html" class="external-link">browseURL</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rWikiPathways/man/getPathwayInfo.html" class="external-link">getPathwayInfo</a></span><span class="op">(</span><span class="va">my.wpids</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/browseURL.html" class="external-link">browseURL</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rWikiPathways/man/getPathwayInfo.html" class="external-link">getPathwayInfo</a></span><span class="op">(</span><span class="va">my.wpids</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span>
<span class="fu"><a href="https://rdrr.io/r/utils/browseURL.html" class="external-link">browseURL</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/rWikiPathways/man/getPathwayInfo.html" class="external-link">getPathwayInfo</a></span><span class="op">(</span><span class="va">my.wpids</span><span class="op">[</span><span class="fl">3</span><span class="op">]</span><span class="op">)</span><span class="op">[</span><span class="fl">2</span><span class="op">]</span><span class="op">)</span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Giulio Caravagna, Luca De Sano, Alex Graudenzi, Daniele Ramazzotti.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link external-link">pkgdown</a> 1.6.1.9001.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
