<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>An introduction to the TRONCO R package • TRONCO</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.7.1/jquery.min.js" integrity="sha512-v2CJ7UaYy4JwqLDIrZUI/4hqeoQieOmAZNXBeQyjo21dadnwR+8ZaIJVT8EE2iyI61OV8e6M8PP2/4hpQINQ/g==" crossorigin="anonymous" referrerpolicy="no-referrer"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="An introduction to the TRONCO R package">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">


    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TRONCO</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.38.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles

    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/f1_introduction.html">An introduction to the TRONCO R package</a>
    </li>
    <li>
      <a href="../articles/f2_loading_data.html">Loading data</a>
    </li>
    <li>
      <a href="../articles/f3_data_visualization.html">Data visualization</a>
    </li>
    <li>
      <a href="../articles/f4_data_manipulation.html">Data manipulation</a>
    </li>
    <li>
      <a href="../articles/f5_model_inference.html">Model inference</a>
    </li>
    <li>
      <a href="../articles/f6_post_reconstruction.html">Post reconstruction</a>
    </li>
    <li>
      <a href="../articles/f7_import_export.html">Import/export from other tools</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BIMIB-DISCo/TRONCO/" class="external-link">
    <span class="fab fa-github fa-lg"></span>

  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->



      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>An introduction to the TRONCO R package</h1>
                        <h4 data-toc-skip class="author">Luca De
Sano</h4>
                        <h4 data-toc-skip class="author">Daniele
Ramazzotti</h4>
                        <h4 data-toc-skip class="author">Giulio
Caravagna</h4>
                        <h4 data-toc-skip class="author">Alex
Graudenxi</h4>
                        <h4 data-toc-skip class="author">Marco
Antoniotti</h4>
            
            <h4 data-toc-skip class="date">October 03, 2024</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BIMIB-DISCo/TRONCO/blob/master/vignettes/f1_introduction.rmd" class="external-link"><code>vignettes/f1_introduction.rmd</code></a></small>
      <div class="hidden name"><code>f1_introduction.rmd</code></div>

    </div>

    
    
<p>The <strong>TRONCO</strong> (TR<em>anslational</em>
ONCO<em>logy</em>) package collects algorithms to infer progression
models via the approach of Suppes-Bayes Causal Network, both from an
ensemble of tumors (cross-sectional samples) and within an individual
patient (multi-region or single-cell samples). The package provides
parallel implementation of algorithms that process binary matrices where
each row represents a tumor sample and each column a single-nucleotide
or a structural variant driving the progression; a 0/1 value models the
absence/presence of that alteration in the sample. The tool can import
data from plain, MAF or GISTIC format files, and can fetch it from the
cBioPortal for cancer genomics. Functions for data manipulation and
visualization are provided, as well as functions to import/export such
data to other bioinformatics tools for, e.g, clustering or detection of
mutually exclusive alterations. Inferred models can be visualized and
tested for their confidence via bootstrap and cross-validation.
<em>TRONCO</em> is used for the implementation of the Pipeline for
Cancer Inference.</p>
<p>In this vignette, we will give an overview of the package by
presenting some of the functions that could be most commonly used to
arrange a data-analysis pipeline, along with their parameters to
customize <em>TRONCO</em>’s functioning. Advanced example case studies
are available at the tool webpage</p>
<div class="section level2">
<h2 id="changelog">Changelog<a class="anchor" aria-label="anchor" href="#changelog"></a>
</h2>
<ul>
<li>[2.31.5] Update vignette.</li>
<li>[2.31.4] Fix error on the usate of <code>order</code> function.</li>
<li>[2.8.1] Minor fix on documentation.</li>
<li>[2.7.7] RNA Seq validation. Random restart on Hill Climbing added to
CAPRI algorithm. Minor fixes to algorithms and error model.</li>
<li>[2.7.3] Development version. Assignment to .GlobalEnv removed.</li>
<li>[2.6.1] Current stable version.</li>
<li>[2.5.3] New algorithms: Edmonds, Gabow, Chow-Liu and Prim. New
scores: PMI, CPMI, MI.</li>
<li>[2.4.3] Bugfix.</li>
<li>[2.4.2] Implements a noise model and finalizes a series of
algorithms reconstructing Suppes-Bayes Causal Network as maximum
spanning trees.</li>
<li>[2.4] New statistics available for model confidence via
cross-validation routines. New algorithms based on Minimum Spanning Tree
extraction.</li>
<li>[2.0] Released in summer 2015 on our GitHUB, replaced the version in
autumn 2015. This version is parallel, includes also the CAPRI
algorithm, supports common GISTIC and MAF input formats, supports TCGA
samples editing and queries to the cBio portal. This version has new
plotting capabilities, and a general from-scratch design. It is not
compatible with previous releases.</li>
<li>[1.0] released in mid 2014, includes CAPRESE algorithm. It is now
outdated and no more maintained;</li>
</ul>
</div>
<div class="section level2">
<h2 id="algorithms-and-useful-links">Algorithms and useful links<a class="anchor" aria-label="anchor" href="#algorithms-and-useful-links"></a>
</h2>
<table class="table">
<colgroup>
<col width="16%">
<col width="34%">
<col width="13%">
<col width="34%">
</colgroup>
<thead><tr class="header">
<th>Acronym</th>
<th>Extended name</th>
<th>App</th>
<th>Reference</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>CAPRESE</td>
<td>Cancer Progression Extraction with Single Edges</td>
<td>Ind</td>
<td><a href="http://www.ncbi.nlm.nih.gov/pubmed/25299648" class="external-link">PLoS ONE,
9(10):e108358, 2014.</a></td>
</tr>
<tr class="even">
<td>CAPRI</td>
<td>Cancer Progression Inference</td>
<td>Ens</td>
<td><a href="http://www.ncbi.nlm.nih.gov/pubmed/25971740" class="external-link">Bioinformatics
31(18), 3016-3016, 2015.</a></td>
</tr>
<tr class="odd">
<td>Edmond</td>
<td>Directed Minimum Spanning Tree with Mutual Information</td>
<td>Ind</td>
<td><a href="https://www.biorxiv.org/content/early/2017/09/04/132183" class="external-link">Publication.</a></td>
</tr>
<tr class="even">
<td>Gabow</td>
<td>Partially Directed Minimum Spanning Tree with Mutual
Information</td>
<td>Ind</td>
<td><a href="https://www.biorxiv.org/content/early/2017/09/04/132183" class="external-link">Publication.</a></td>
</tr>
<tr class="odd">
<td>Chow Liu</td>
<td>Undirected Minimum Spanning Tree with Likelihood-Fit</td>
<td>Ind</td>
<td><a href="https://www.biorxiv.org/content/early/2017/09/04/132183" class="external-link">Publication.</a></td>
</tr>
<tr class="even">
<td>Prim</td>
<td>Undirected Minimum Spanning Tree with Mutual Information</td>
<td>Ind</td>
<td><a href="https://www.biorxiv.org/content/early/2017/09/04/132183" class="external-link">Publication.</a></td>
</tr>
</tbody>
</table>
<p><strong>Legend</strong></p>
<ul>
<li>Ens.: ensemble-level with cross-sectional data;</li>
<li>Ind.: individual-level with single-cell or multi-region data.</li>
</ul>
</div>
<div class="section level2">
<h2 id="external-links-to-resources-related-to-tronco">External links to resources related to TRONCO<a class="anchor" aria-label="anchor" href="#external-links-to-resources-related-to-tronco"></a>
</h2>
<ul>
<li>TRONCO was introduced in <a href="https://academic.oup.com/bioinformatics/article/32/12/1911/1743307" class="external-link">De
Sano, Luca, et al. “TRONCO: an R package for the inference of cancer
progression models from heterogeneous genomic data.” Bioinformatics
32.12 (2016): 1911-1913</a>.</li>
<li>TRONCO since version <em>2.3</em> is used to implement the
<strong>Pipeline For Cancer Inference (PiCnIc)</strong> described in <a href="http://www.pnas.org/content/113/28/E4025" class="external-link">Caravagna, Giulio, et
al. “Algorithmic methods to infer the evolutionary trajectories in
cancer progression.” Proceedings of the National Academy of Sciences
113.28 (2016): E4025-E4034</a>.</li>
<li>Case studies featuring Atypical Chronic Myeloid Leukemia, Colorectal
Cancer, Clear Cell Renal Cell Carcinoma and others are available at the
tool <a href="https://sites.google.com/site/troncopackage/" class="external-link">webpage</a>.
Code for replication of each of those study is made available through <a href="https://github.com/BIMIB-DISCo" class="external-link">Bioinformatics Milano-Bicocca’s
Github</a>.</li>
</ul>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

      </div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Giulio Caravagna, Luca De Sano, Alex Graudenzi, Daniele Ramazzotti.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.1.9000.</p>
</div>

      </footer>
</div>






  </body>
</html>
