<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Loading data • TRONCO</title>
<!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous">
<script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css">
<script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../pkgdown.js"></script><meta property="og:title" content="Loading data">
<meta property="og:description" content="TRONCO">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body data-spy="scroll" data-target="#toc">
    

    <div class="container template-article">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">TRONCO</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">2.32.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../reference/index.html">Reference</a>
</li>
<li class="dropdown">
  <a href="#" class="dropdown-toggle" data-toggle="dropdown" role="button" data-bs-toggle="dropdown" aria-expanded="false">
    Articles
     
    <span class="caret"></span>
  </a>
  <ul class="dropdown-menu" role="menu">
<li>
      <a href="../articles/f1_introduction.html">An introduction to the TRONCO R package</a>
    </li>
    <li>
      <a href="../articles/f2_loading_data.html">Loading data</a>
    </li>
    <li>
      <a href="../articles/f3_data_visualization.html">Data visualization</a>
    </li>
    <li>
      <a href="../articles/f4_data_manipulation.html">Data manipulation</a>
    </li>
    <li>
      <a href="../articles/f5_model_inference.html">Model inference</a>
    </li>
    <li>
      <a href="../articles/f6_post_reconstruction.html">Post reconstruction</a>
    </li>
    <li>
      <a href="../articles/f7_import_export.html">Import/export from other tools</a>
    </li>
  </ul>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/BIMIB-DISCo/TRONCO/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header toc-ignore">
      <h1 data-toc-skip>Loading data</h1>
                        <h4 data-toc-skip class="author">Luca De
Sano</h4>
                        <h4 data-toc-skip class="author">Daniele
Ramazzotti</h4>
                        <h4 data-toc-skip class="author">Giulio
Caravagna</h4>
                        <h4 data-toc-skip class="author">Alex
Graudenxi</h4>
                        <h4 data-toc-skip class="author">Marco
Antoniotti</h4>
            
            <h4 data-toc-skip class="date">April 29, 2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/BIMIB-DISCo/TRONCO/blob/HEAD/vignettes/f2_loading_data.Rmd" class="external-link"><code>vignettes/f2_loading_data.Rmd</code></a></small>
      <div class="hidden name"><code>f2_loading_data.Rmd</code></div>

    </div>

    
    
<div class="section level2">
<h2 id="preliminaries">Preliminaries<a class="anchor" aria-label="anchor" href="#preliminaries"></a>
</h2>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://sites.google.com/site/troncopackage/" class="external-link">TRONCO</a></span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">aCML</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">crc_maf</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">crc_gistic</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="va">crc_plain</span><span class="op">)</span></span></code></pre></div>
<p>TRONCO transforms input data in a sort of database-alike format,
where three main fields are presente: <em>genotypes</em> which contains
the genomic signatures of the input samples, <em>annotations</em> which
provides an index to the events present in the data and <em>types</em>,
a field mapping type of events (e.g., mutations, CNAs, etc.) to colors
for display visualization. Other annotations are generated when a
dataset is augmented with some metadata. A TRONCO object shall be edited
by using TRONCO functions, to avoid to create inconsistencies in its
internal representation. Function <code>is.compliant</code> can be used
to test if a TRONCO object is consistent; the function is called by any
TRONCO function before returning a modified object, so to ensure that
consistency is preserved – <code>is.compliant</code> will raise an error
if this is not the case.</p>
<p>TRONCO supports the import of data from 3 formats. The Mutation
Annotation Format (<em>MAF</em>) is a tab-delimited file containing
somatic and/or germline mutation annotations; the <em>GISTIC</em> format
for copy number alterations as defined by TCGA and a custom boolean
matrix format where the user can directly specify the mutational
profiles to be importend. Through some data included in the package we
will show how to load your datasets in TRONCO.</p>
<ul>
<li>
<strong>aCML</strong> a TRONCO object that represents the
<em>atypical Chronic Myeloid Leukemia</em> dataset by Piazza <em>et
al.</em> (Nat. Gen. 2013 45(1):18-24).</li>
<li>
<strong>crc_maf</strong> a shortened version of the <em>colorectal
cancer mutation data</em> made available by the TCGA consortium within
the COADREAD. See <a href="https://tcga-data.nci.nih.gov/docs/publications/coadread_2012/" class="external-link">this
link</a> and our PicNiC case study for the real analysis of such
data.</li>
<li>
<strong>crc_gistic</strong> from the same TCGA project, we also
provide a shortened version of the focal CNAs in the GISTIC format where
1 represents a low level gain, 2 a high level gain, -1 a heterozygous
loss of a gene and -2 its homozygous loss.</li>
<li>
<strong>crc_plain</strong> a custom boolean matrix where rows are
samples, and columns represent events – in this case alterations in a
certain gene. Notice with this format one could also custom types of
alterations, for instance wider chromosomal aberrations or, in
principle, epigenetic states (over-expression, methylated regions, etc.)
that are persistent across tumor evolution.</li>
</ul>
<p>Whatever is dataset created as explained in the next sections, it can
be annotated by adding a mnemonic description of the data, which will be
used as plot titles when possible. Function
<code>annotate.description</code> raises a warning if the dataset was
previously annotated.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">aCML</span> <span class="op">=</span> <span class="fu"><a href="../reference/annotate.description.html">annotate.description</a></span><span class="op">(</span><span class="va">aCML</span>, <span class="st">'aCML data (Bioinf.)'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="mutations-annotated-in-a-maf-format">Mutations annotated in a MAF format<a class="anchor" aria-label="anchor" href="#mutations-annotated-in-a-maf-format"></a>
</h2>
<p>We use the function <code>import.MAF</code> to import a dataset in
MAF format, in this case the following TCGA dataset</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">crc_maf</span><span class="op">[</span>, <span class="fl">1</span><span class="op">:</span><span class="fl">10</span><span class="op">]</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      Hugo_Symbol Entrez_Gene_Id       Center NCBI_Build Chromosome</span></span>
<span><span class="co">## 27          TP53           7157 hgsc.bcm.edu         36         17</span></span>
<span><span class="co">## 246        FBXW7          55294 hgsc.bcm.edu         36          4</span></span>
<span><span class="co">## 623          APC            324 hgsc.bcm.edu         36          5</span></span>
<span><span class="co">## 649         TP53           7157 hgsc.bcm.edu         36         17</span></span>
<span><span class="co">## 928        FBXW7          55294 hgsc.bcm.edu         36          4</span></span>
<span><span class="co">## 1390        TP53           7157 hgsc.bcm.edu         36         17</span></span>
<span><span class="co">##      Start_position End_position Strand Variant_Classification Variant_Type</span></span>
<span><span class="co">## 27          7519131      7519131      +      Missense_Mutation          SNP</span></span>
<span><span class="co">## 246       153466817    153466817      +      Nonsense_Mutation          SNP</span></span>
<span><span class="co">## 623       112192485    112192485      +      Nonsense_Mutation          SNP</span></span>
<span><span class="co">## 649         7518937      7518937      +      Nonsense_Mutation          SNP</span></span>
<span><span class="co">## 928       153468834    153468834      +      Missense_Mutation          SNP</span></span>
<span><span class="co">## 1390        7517864      7517864      +      Missense_Mutation          SNP</span></span></code></pre>
<p>A default importation is done without adding parameters to
<code>import.MAF</code>. In this case, all mutations per gene will be
considered equivalent, regardless of the type that is annotated in the
MAF. Also, all genes will be imported, and all samples.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset_maf</span> <span class="op">=</span> <span class="fu"><a href="../reference/import.MAF.html">import.MAF</a></span><span class="op">(</span><span class="va">crc_maf</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## *** Importing from dataframe</span></span>
<span><span class="co">## Loading MAF dataframe ...DONE</span></span>
<span><span class="co">## *** Mutations names: using Hugo_Symbol</span></span>
<span><span class="co">## *** Using full MAF: #entries  17 </span></span>
<span><span class="co">## *** MAF report: TCGA=TRUE</span></span>
<span><span class="co">## Type of annotated mutations: </span></span>
<span><span class="co">## [1] "Missense_Mutation" "Nonsense_Mutation"</span></span>
<span><span class="co">## *** [merge.mutation.types = T] Mutations will be merged and annotated as 'Mutation'</span></span>
<span><span class="co">## Number of samples: 9 </span></span>
<span><span class="co">## [TCGA = TRUE] Number of TCGA patients: 9 </span></span>
<span><span class="co">## Number of annotated mutations: 17 </span></span>
<span><span class="co">## Mutations annotated with "Valid" flag (%): 71 </span></span>
<span><span class="co">## Number of genes (Hugo_Symbol): 6 </span></span>
<span><span class="co">## Starting conversion from MAF to 0/1 mutation profiles (1 = mutation) :9 x 6 </span></span>
<span><span class="co">## .................</span></span>
<span><span class="co">## Starting conversion from MAF to TRONCO data type.</span></span></code></pre>
<p>In the above case – where we see that mutations are annotated as
<code>Missense_Mutation</code> or <code>Nonsense_Mutation</code>, if a
gene in a sample has both, these will be merged to a unique
<code>Mutation</code> type. In this case a pair gene name with
<code>Mutation</code> will be what we call an <em>event</em> in our
dataset – e.g., APC Mutation. If one would like to have two distinct
events in the dataset, i.e., APC <code>Missense_Mutation</code> and APC
<code>Nonsense_Mutation</code>, parameter
<code>merge.mutation.types</code> should be set to false in the call to
<code>import.MAF</code>.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset_maf</span> <span class="op">=</span> <span class="fu"><a href="../reference/import.MAF.html">import.MAF</a></span><span class="op">(</span><span class="va">crc_maf</span>, merge.mutation.types <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## *** Importing from dataframe</span></span>
<span><span class="co">## Loading MAF dataframe ...DONE</span></span>
<span><span class="co">## *** Mutations names: using Hugo_Symbol</span></span>
<span><span class="co">## *** Using full MAF: #entries  17 </span></span>
<span><span class="co">## *** MAF report: TCGA=TRUE</span></span>
<span><span class="co">## Type of annotated mutations: </span></span>
<span><span class="co">## [1] "Missense_Mutation" "Nonsense_Mutation"</span></span>
<span><span class="co">## *** [merge.mutation.types = F] Mutations will be distinguished by type</span></span>
<span><span class="co">## Number of samples: 9 </span></span>
<span><span class="co">## [TCGA = TRUE] Number of TCGA patients: 9 </span></span>
<span><span class="co">## Number of annotated mutations: 17 </span></span>
<span><span class="co">## Mutations annotated with "Valid" flag (%): 71 </span></span>
<span><span class="co">## Number of genes (Hugo_Symbol): 6 </span></span>
<span><span class="co">## Starting conversion from MAF to 0/1 mutation profiles (1 = mutation) :</span></span>
<span><span class="co">## .................</span></span></code></pre>
<p>Sometimes, we might want to filter out some of the entries in a MAF –
maybe restricting the type of genes, mutations or sample that we want to
process. If one defines <code>filter.fun</code> as a function that
returns <code>TRUE</code> only for those entries which shall be
considered, he gets a filter process which is applied to each row of the
MAF file prior to transforming that into a TRONCO dataset. In this
example we select only mutations annotated to APC – we access that
through the <em>Hugo_Symbol</em> flag of a MAF.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset_maf</span> <span class="op">=</span> <span class="fu"><a href="../reference/import.MAF.html">import.MAF</a></span><span class="op">(</span><span class="va">crc_maf</span>, filter.fun <span class="op">=</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span><span class="op">)</span><span class="op">{</span> <span class="va">x</span><span class="op">[</span><span class="st">'Hugo_Symbol'</span><span class="op">]</span> <span class="op">==</span> <span class="st">'APC'</span><span class="op">}</span> <span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## *** Importing from dataframe</span></span>
<span><span class="co">## Loading MAF dataframe ...DONE</span></span>
<span><span class="co">## *** Mutations names: using Hugo_Symbol</span></span>
<span><span class="co">## *** Filtering full MAF: #entries  17 </span></span>
<span><span class="co">## *** Using reduced MAF: #entries  3 </span></span>
<span><span class="co">## *** MAF report: TCGA=TRUE</span></span>
<span><span class="co">## Type of annotated mutations: </span></span>
<span><span class="co">## [1] "Nonsense_Mutation"</span></span>
<span><span class="co">## *** [merge.mutation.types = T] Mutations will be merged and annotated as 'Mutation'</span></span>
<span><span class="co">## Number of samples: 3 </span></span>
<span><span class="co">## [TCGA = TRUE] Number of TCGA patients: 3 </span></span>
<span><span class="co">## Number of annotated mutations: 3 </span></span>
<span><span class="co">## Mutations annotated with "Valid" flag (%): 33 </span></span>
<span><span class="co">## Number of genes (Hugo_Symbol): 1 </span></span>
<span><span class="co">## Starting conversion from MAF to 0/1 mutation profiles (1 = mutation) :3 x 1 </span></span>
<span><span class="co">## ...</span></span>
<span><span class="co">## Starting conversion from MAF to TRONCO data type.</span></span></code></pre>
<p>It is also sometimes convenient – especially when working with data
collected from a single individual patient – to distinguish the type of
mutations and their position in a gene, or if they are somehow annotated
to COSMIC or other databases. For instance, we might want to want to use
the <code>MA.protein.change</code> annotation in the MAF file to get
composite names such as TP53.R175H, TP53.R213, TP53.R267W etc. This can
be done by setting {} to have the relevant name of the MAF
annotation</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset_maf</span> <span class="op">=</span> <span class="fu"><a href="../reference/import.MAF.html">import.MAF</a></span><span class="op">(</span><span class="va">crc_maf</span>, </span>
<span>    merge.mutation.types <span class="op">=</span> <span class="cn">FALSE</span>, </span>
<span>    paste.to.Hugo_Symbol <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'MA.protein.change'</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## *** Importing from dataframe</span></span>
<span><span class="co">## Loading MAF dataframe ...DONE</span></span>
<span><span class="co">## *** Mutations names: augmenting Hugo_Symbol with values:  MA.protein.change </span></span>
<span><span class="co">## *** Using full MAF: #entries  17 </span></span>
<span><span class="co">## *** MAF report: TCGA=TRUE</span></span>
<span><span class="co">## Type of annotated mutations: </span></span>
<span><span class="co">## [1] "Missense_Mutation" "Nonsense_Mutation"</span></span>
<span><span class="co">## *** [merge.mutation.types = F] Mutations will be distinguished by type</span></span>
<span><span class="co">## Number of samples: 9 </span></span>
<span><span class="co">## [TCGA = TRUE] Number of TCGA patients: 9 </span></span>
<span><span class="co">## Number of annotated mutations: 17 </span></span>
<span><span class="co">## Mutations annotated with "Valid" flag (%): 71 </span></span>
<span><span class="co">## Number of genes (Hugo_Symbol): 16 </span></span>
<span><span class="co">## Starting conversion from MAF to 0/1 mutation profiles (1 = mutation) :</span></span>
<span><span class="co">## .................</span></span></code></pre>
<p>TRONCO supports custom MAF files, where possibly not all the standard
annotations are present, via <code>irregular = TRUE</code>.</p>
</div>
<div class="section level2">
<h2 id="copy-number-variants-annotated-in-the-gistic-format">Copy Number Variants annotated in the GISTIC format<a class="anchor" aria-label="anchor" href="#copy-number-variants-annotated-in-the-gistic-format"></a>
</h2>
<p>We use the function <code>import.GISTIC</code> to import a dataset in
GISTIC format, in this case from</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc_gistic</span></span></code></pre></div>
<pre><code><span><span class="co">##              NRAS CTNNB1 FBXW7 APC KRAS TP53</span></span>
<span><span class="co">## TCGA-A6-2670   -1      0     0  -1    1   -1</span></span>
<span><span class="co">## TCGA-A6-2672    0      0     0   0    0    0</span></span>
<span><span class="co">## TCGA-A6-2674    0      0     0   0    0    0</span></span>
<span><span class="co">## TCGA-A6-2676    0      0     0   0    0    0</span></span>
<span><span class="co">## TCGA-A6-2677    0      0     0   0    0   -1</span></span>
<span><span class="co">## TCGA-A6-2678    0      0     0   0    0   -1</span></span>
<span><span class="co">## TCGA-A6-2683   -1     -1    -1  -1    0   -1</span></span>
<span><span class="co">## TCGA-A6-3807    0     -1     0   0   -1   -1</span></span>
<span><span class="co">## TCGA-AA-3516    0      0     0   0    0    0</span></span></code></pre>
<p>In its default execution all the data annotated in the file is
imported. But in principle it is possible to avoid to import some genes
or samples; in this case it is sufficient to use parameters
<code>filter.genes</code> and <code>filter.samples</code> for this
function.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset_gistic</span> <span class="op">=</span> <span class="fu"><a href="../reference/import.GISTIC.html">import.GISTIC</a></span><span class="op">(</span><span class="va">crc_gistic</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## *** Using full GISTIC: #dim  9  x  6 </span></span>
<span><span class="co">## *** GISTIC input format conversion started.</span></span>
<span><span class="co">## Converting input data to character for import speedup.</span></span>
<span><span class="co">## Creating  24 events for 6 genes </span></span>
<span><span class="co">## Extracting "Homozygous Loss" events (GISTIC = -2) </span></span>
<span><span class="co">## Extracting "Heterozygous Loss" events (GISTIC = -1) </span></span>
<span><span class="co">## Extracting "Low-level Gain" events (GISTIC = +1) </span></span>
<span><span class="co">## Extracting "High-level Gain" events (GISTIC = +2) </span></span>
<span><span class="co">## Transforming events in TRONCO data types ..... </span></span>
<span><span class="co">## *** Binding events for 4 datasets.</span></span>
<span><span class="co">## *** Data extracted, returning only events observed in at least one sample </span></span>
<span><span class="co">##  Number of events: n = 7 </span></span>
<span><span class="co">##  Number of genes: |G| = 6 </span></span>
<span><span class="co">##  Number of samples: m = 9</span></span></code></pre>
</div>
<div class="section level2">
<h2 id="custom-alterations-annotated-in-a-boolean-matrix">Custom alterations annotated in a boolean matrix<a class="anchor" aria-label="anchor" href="#custom-alterations-annotated-in-a-boolean-matrix"></a>
</h2>
<p>One can annotate its custom type of alterations in a boolean matrix
such as <code>crc_plain</code></p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">crc_plain</span></span></code></pre></div>
<pre><code><span><span class="co">##                 TP53 FBXW7 APC CTNNB1 NRAS KRAS</span></span>
<span><span class="co">## TCGA-AA-3517-01    1     0   0      0    0    0</span></span>
<span><span class="co">## TCGA-AA-3518-01    0     1   0      0    0    0</span></span>
<span><span class="co">## TCGA-AA-3519-01    1     0   1      0    0    0</span></span>
<span><span class="co">## TCGA-AA-3520-01    1     0   0      0    0    1</span></span>
<span><span class="co">## TCGA-AA-3521-01    0     0   1      0    0    1</span></span></code></pre>
<p>In this case, function <code>import.genotypes</code> will convert the
matrix to a TRONCO object where events’ names and samples codes will be
set from column and row names of the matrix. If this is not possible,
these will be generated from templates. By default, the
<code>event.type</code> is set to <code>variant</code> but one can
specify a custom name for the alteration that is reported in the
matrix</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">dataset_plain</span> <span class="op">=</span> <span class="fu"><a href="../reference/import.genotypes.html">import.genotypes</a></span><span class="op">(</span><span class="va">crc_plain</span>, event.type<span class="op">=</span><span class="st">'myVariant'</span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="downloading-data-from-the-cbio-portal-for-cancer-genomics">Downloading data from the cBio portal for cancer genomics<a class="anchor" aria-label="anchor" href="#downloading-data-from-the-cbio-portal-for-cancer-genomics"></a>
</h2>
<p>TRONCO uses the R interface to cBio to query data from the portal.
All type of data can be downloaded from the portal, which includes
MAF/GISTIC data for a lot of different cancer studies. An example of
interaction with the portal is archived at the tool’s webpage.</p>
<p>Here, we show how to download lung cancer data somatic mutations for
genes TP53, KRAS and PIK3CA, from the lung cancer project run by TCGA,
which is archived as <em>luad_tcga_pub</em> at cBio. If some of the
parameters to <code>cbio.query</code> are missing the function will
become interactive by showing a list of possible data available at the
portal.</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">data</span> <span class="op">=</span> <span class="fu">cbio.query</span><span class="op">(</span></span>
<span>    genes<span class="op">=</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">'TP53'</span>, <span class="st">'KRAS'</span>, <span class="st">'PIK3CA'</span><span class="op">)</span>, </span>
<span>    cbio.study <span class="op">=</span> <span class="st">'luad_tcga_pub'</span>,</span>
<span>    cbio.dataset <span class="op">=</span> <span class="st">'luad_tcga_pub_cnaseq'</span>, </span>
<span>    cbio.profile <span class="op">=</span> <span class="st">'luad_tcga_pub_mutations'</span><span class="op">)</span></span></code></pre></div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">

        <nav id="toc" data-toggle="toc"><h2 data-toc-skip>Contents</h2>
    </nav>
</div>

</div>



      <footer><div class="copyright">
  <p></p>
<p>Developed by Giulio Caravagna, Luca De Sano, Alex Graudenzi, Daniele Ramazzotti.</p>
</div>

<div class="pkgdown">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.9000.</p>
</div>

      </footer>
</div>

  


  

  </body>
</html>
